lib/cache/bot_instance.go

```diff
@@ -158,33 +142,50 @@ func (c *Cache) ListBotInstances(ctx context.Context, botName string, pageSize i
 	return out, next, trace.Wrap(err)
 }
 
-func matchBotInstance(b *machineidv1.BotInstance, botName string, search string) bool {
-	// If updating this, ensure it's consistent with the upstream search logic in `lib/services/local/bot_instance.go`.
+func keyForNameIndex(botInstance *machineidv1.BotInstance) string {
+	return makeNameIndexKey(
+		botInstance.GetSpec().GetBotName(),
+		botInstance.GetMetadata().GetName(),
+	)
+}
 
-	if botName != "" && b.Spec.BotName != botName {
-		return false
-	}
+func makeNameIndexKey(botName string, instanceID string) string {
+	return botName + "/" + instanceID
+}
 
-	if search == "" {
-		return true
+func keyForActiveAtIndex(botInstance *machineidv1.BotInstance) string {
+	heartbeat := services.GetBotInstanceLatestHeartbeat(botInstance)
+	recordedAt := heartbeat.GetRecordedAt().AsTime()
+	return recordedAt.Format(time.RFC3339) + "/" + botInstance.GetMetadata().GetName()
+}
+
+func keyForVersionIndex(botInstance *machineidv1.BotInstance) string {
+	version := "000000.000000.000000"
+	heartbeat := services.GetBotInstanceLatestHeartbeat(botInstance)
+	if heartbeat == nil {
+		return version + "/" + botInstance.GetMetadata().GetName()
 	}
 
-	latestHeartbeats := b.GetStatus().GetLatestHeartbeats()
-	heartbeat := b.Status.InitialHeartbeat // Use initial heartbeat as a fallback
-	if len(latestHeartbeats) > 0 {
-		heartbeat = latestHeartbeats[len(latestHeartbeats)-1]
+	sv, err := semver.NewVersion(heartbeat.GetVersion())
+	if err != nil {
+		return version + "/" + botInstance.GetMetadata().GetName()
 	}
 
-	values := []string{
-		b.Spec.BotName,
-		b.Spec.InstanceId,
+	version = fmt.Sprintf("%06d.%06d.%06d", sv.Major, sv.Minor, sv.Patch)
+	if sv.PreRelease != "" {
+		version = version + "-" + string(sv.PreRelease)
+	}
+	if sv.Metadata != "" {
+		version = version + "+" + sv.Metadata
 	}
```

Comment:
I don't know how much we care, but the ordering of semver is more complicated than just zero-padding the triplet of version numbers (`1.2.3` is greater than `1.2.3-a`, `1.1.1-beta.4` is smaller than `1.1.1-beta.32`).